packages/libs/error-stack/src/ext/iter.rs

```diff
@@ -0,0 +1,353 @@
+    /// [`Iterator::next`]
+    ///
+    /// Applies [`Report::change_context`] to every [`Err`] variant in the iterator. For more
+    /// information, see the documentation for [`Report::change_context`].
+    ///
+    /// [`Report`]: crate::Report
+    /// [`Report::change_context_lazy`]: crate::Report::change_context_lazy
+    /// [`poll`]: Future::poll
+    fn change_context_lazy<C, F>(self, context: F) -> IteratorWithLazyContext<Self, F>
+    where
+        C: Context,
+        F: Fn() -> C;
+}
+
+impl<I: Iterator> IteratorExt for I
+where
+    I::Item: ResultExt,
+{
+    #[track_caller]
+    fn attach<A>(self, attachment: A) -> IteratorWithAttachment<Self, A> {
+        IteratorWithAttachment {
+            iterator: self,
+            context: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_lazy<A, F>(self, attachment: F) -> IteratorWithLazyAttachment<Self, F>
+    where
+        A: Send + Sync + 'static,
+        F: Fn() -> A,
+    {
+        IteratorWithLazyAttachment {
+            iterator: self,
+            context: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_printable<A>(self, attachment: A) -> IteratorWithPrintableAttachment<Self, A>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+    {
+        IteratorWithPrintableAttachment {
+            iterator: self,
+            context: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn attach_printable_lazy<A, F>(
+        self,
+        attachment: F,
+    ) -> IteratorWithLazyPrintableAttachment<Self, F>
+    where
+        A: Display + Debug + Send + Sync + 'static,
+        F: Fn() -> A,
+    {
+        IteratorWithLazyPrintableAttachment {
+            iterator: self,
+            context: attachment,
+        }
+    }
+
+    #[track_caller]
+    fn change_context<C>(self, context: C) -> IteratorWithContext<Self, C>
+    where
+        C: Context,
+    {
+        IteratorWithContext {
+            iterator: self,
+            context,
+        }
+    }
+
+    #[track_caller]
+    fn change_context_lazy<C, F>(self, context: F) -> IteratorWithLazyContext<Self, F>
+    where
+        C: Context,
+        F: Fn() -> C,
+    {
+        IteratorWithLazyContext {
+            iterator: self,
+            context,
+        }
+    }
+}
+
+pub struct IteratorWithAttachment<I, A> {
+    iterator: I,
+    context: A,
+}
+
+impl<I, A> Iterator for IteratorWithAttachment<I, A>
+where
+    I: Iterator,
+    I::Item: ResultExt,
+    A: Send + Sync + 'static + Clone,
+{
+    type Item = I::Item;
+
+    #[track_caller]
+    fn next(&mut self) -> Option<Self::Item> {
+        Some(self.iterator.next()?.attach_lazy(|| self.context.clone()))
+    }
+}
+
+pub struct IteratorWithLazyAttachment<I, A> {
+    iterator: I,
+    context: A,
+}
+
+impl<I, F, A> Iterator for IteratorWithLazyAttachment<I, F>
+where
+    I: Iterator,
+    I::Item: ResultExt,
+    F: Fn() -> A + Clone + Send + Sync + 'static,
```

Comment:
The closure does not need the bounds
```suggestion
    F: Fn() -> A,
```